### PR TITLE
ScaleRL: Add CISPO Loss

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?
CISPO Loss was first introduced in the Minimax-M1 [paper](https://arxiv.org/pdf/2506.13585v1), the [ScaleRL](https://arxiv.org/pdf/2510.13786) paper subsequently showed that CISPO Loss scales the best in terms of performance and efficiency as models are trained for longer.

image

Where rho is

image

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1818,17 +1818,21 @@ def _compute_loss(self, model, inputs):
             )
         # From here, log_importance_weights (and all subsequent tensors, coef_1, coef_2, etc.) shape depends on
         # importance_sampling_level: "token" level: (B, T); "sequence" level: (B, 1)
+        if self.loss_type == "cispo":
+            coef_1 = torch.clamp(torch.exp(log_importance_weights), max=self.epsilon_high).detach()
+            per_token_loss = coef_1 * advantages.unsqueeze(1) * per_token_logps
```
[the remark this answers is not in the document]
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: Nice catch copilot!

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1871,23 +1875,30 @@ def masked_batch_mean(x):
...
+            gathered_low_clip = self.accelerator.gather(low_clip)
+            self._metrics[mode]["clip_ratio/low_mean"].append(gathered_low_clip.nanmean().item())
+            self._metrics[mode]["clip_ratio/low_min"].append(nanmin(gathered_low_clip).item())
+            gathered_high_clip = self.accelerator.gather(high_clip)
+            self._metrics[mode]["clip_ratio/high_mean"].append(gathered_high_clip.nanmean().item())
+            self._metrics[mode]["clip_ratio/high_max"].append(nanmax(gathered_high_clip).item())
+            gathered_clip_ratio = self.accelerator.gather(clip_ratio)
+            self._metrics[mode]["clip_ratio/region_mean"].append(gathered_clip_ratio.nanmean().item())
+        else:
+            is_cispo_clipped = (log_importance_weights > 1 + self.epsilon_high) & (advantages.unsqueeze(1) > 0)
```
[the remark this answers is not in the document]
Comment (pramodith) [reply trl/trainer/grpo_trainer.py]: Good catch but the wrong suggestion

File: docs/source/paper_index.md [hunk 3]
```diff
@@ -232,6 +232,28 @@ trainer = PAPOTrainer(
 )
 ```
 
+### The Art of Scaling Reinforcement Learning
+
+**📜 Paper**: https://huggingface.co/papers/2510.13786
+
+A systematic study that defines a framework for analyzing and predicting reinforcement learning scaling in large language models, identifies key design choices that affect compute efficiency and propose a best-practice recipe called ScaleRL.
+
+You can partially reproduce the ScaleRL recipe using the `GRPOTrainer` with the following configs:
```
Comment (qgallouedec) [inline docs/source/paper_index.md]:
```suggestion
You can partially reproduce the ScaleRL recipe using the [`GRPOTrainer`] with the following configs:
```

File: docs/source/paper_index.md [hunk 4]
```diff
@@ -232,6 +232,28 @@ trainer = PAPOTrainer(
...
+
+```python
+from trl import GRPOConfig
+
+config = GRPOConfig(
+    loss_type="cispo",
+    epsilon_high=5,
```
Comment (qgallouedec) [inline docs/source/paper_index.md]: nit, it's a float, so I think it's better:

```suggestion
    epsilon_high=5.0,
```

File: trl/trainer/grpo_trainer.py [hunk 5]
```diff
@@ -1871,23 +1877,30 @@ def masked_batch_mean(x):
...
-
-        gathered_low_clip = self.accelerator.gather(low_clip)
-        self._metrics[mode]["clip_ratio/low_mean"].append(gathered_low_clip.nanmean().item())
-        self._metrics[mode]["clip_ratio/low_min"].append(nanmin(gathered_low_clip).item())
-        gathered_high_clip = self.accelerator.gather(high_clip)
-        self._metrics[mode]["clip_ratio/high_mean"].append(gathered_high_clip.nanmean().item())
-        self._metrics[mode]["clip_ratio/high_max"].append(nanmax(gathered_high_clip).item())
-        gathered_clip_ratio = self.accelerator.gather(clip_ratio)
-        self._metrics[mode]["clip_ratio/region_mean"].append(gathered_clip_ratio.nanmean().item())
+        if self.loss_type != "cispo":
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1880]: nit, again (explicit better than implicit

```suggestion
        if self.loss_type in ["grpo", "bnpo", "dr_grpo", "dapo"]:
```

File: trl/trainer/grpo_trainer.py [hunk 6]
```diff
@@ -1871,23 +1877,30 @@ def masked_batch_mean(x):
...
+
+            gathered_low_clip = self.accelerator.gather(low_clip)
+            self._metrics[mode]["clip_ratio/low_mean"].append(gathered_low_clip.nanmean().item())
+            self._metrics[mode]["clip_ratio/low_min"].append(nanmin(gathered_low_clip).item())
+            gathered_high_clip = self.accelerator.gather(high_clip)
+            self._metrics[mode]["clip_ratio/high_mean"].append(gathered_high_clip.nanmean().item())
+            self._metrics[mode]["clip_ratio/high_max"].append(nanmax(gathered_high_clip).item())
+            gathered_clip_ratio = self.accelerator.gather(clip_ratio)
+            self._metrics[mode]["clip_ratio/region_mean"].append(gathered_clip_ratio.nanmean().item())
+        else:
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py]:
```suggestion
        elif loss_type == "cispo":
```

File: trl/trainer/grpo_config.py [hunk 7]
```diff
@@ -201,6 +201,9 @@ class GRPOConfig(TrainingArguments):
               batch. Note that normalization is performed over the local batch only, so results may slightly vary
               depending on the local batch size, despite a constant effective batch size. When using
               `per_device_train_batch_size==1`, the loss is equivalent to the GRPO loss.
+            - `"cispo"`: Clips the importance sampling weights instead of the advantage scaled importance weights. The clipped weights
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]: can you just make sure that the line length <= 120

File: trl/trainer/grpo_config.py [hunk 8]
```diff
@@ -201,6 +201,9 @@ class GRPOConfig(TrainingArguments):
...
+              are then multiplied with the advantages and policy model's log probs. Individual token losses are aggregated by
+              normalizing with the number of active tokens in the global accumulated batch. This method was introduced in the MiniMax-M1 paper.
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
              normalizing with the number of active tokens in the global accumulated batch. This method was introduced in the [MiniMax-M1 paper](https://huggingface.co/papers/2506.13585).
```

File: trl/trainer/grpo_trainer.py [hunk 9]
```diff
@@ -1816,19 +1816,25 @@ def _compute_loss(self, model, inputs):
                 f"Unknown importance sampling level: {self.importance_sampling_level}. Possible values are 'token' "
                 "and 'sequence'."
             )
+
+        coef_1 = torch.exp(log_importance_weights)
+
         # From here, log_importance_weights (and all subsequent tensors, coef_1, coef_2, etc.) shape depends on
         # importance_sampling_level: "token" level: (B, T); "sequence" level: (B, 1)
+        if self.loss_type == "cispo":
+            clamped_ratios = torch.clamp(coef_1, max=self.epsilon_high).detach()
```
Comment (qgallouedec) [inline trl/trainer/grpo_trainer.py:1825]: maybe in the documentation of `epsilon_high` we can mention that this is the value used for epsilon_max when used with CISPO loss. and that the paper recommends =5.0

File: docs/source/paper_index.md (hunk 3 above)
Comment (qgallouedec) [inline docs/source/paper_index.md]: not needed for this PR, but it would be neat to have a list of what's supported and what's not. If useful, these are some light reading notes

```
baseline: GRPO beta=0 and asymetric clipping

scaling low applies after ∼1.5k GPU hours

3 stages:
1. Ablation: Test individual design choices on the baseline (3.5k–4k GPU-hours) to identify stable ones.
2. LOO Experiments: Combine stable choices into *ScaleRL* and run 16k GPU-hour leave-one-out tests to assess predictability (fit first 8k, extrapolate rest).
3. Scaling Demonstration: Validate ScaleRL predictability on larger, more complex setups (bigger batches, MoE models, multitask, longer sequences).

### Asynchronous RL Setup

**PPO-off-policy-k** is equivalent to `steps_per_generation` in TRL. -> faster convergence `steps_per_generation=8` than `steps_per_generation=1`
**PipelineRL-k** acheives true asyncronisity by allowing weight updates during rollouts. -> faster convergence than **PPO-off-policy-k**

Note that the asymptotic performance is the same for these methods.

### Algorithmic Choices

* loss type: DAPO/GSPO/CISPO: CISPO > GSPO >> DAPO (not yet in TRL, but PR opened by a contributor)
* precision fixes: fp32 for lm_head substantially improves final performance (`cast_lm_head_to_fp32=True` in TRL)
* loss aggregation: (GRPO vs DAPO vs DrGRPO): DAPO-style works best (`loss_type="dapo"` in TRL)
* advantage normalization (prompt-level vs batch-level vs none): equivalent, choose batch-level ( `scale_rewards="batch"` in TRL)
* batch definition: zero-Variance filtering gives better final performance (not implemented in TRL!)
* data curriculum: filter "too easy" prompt are filtered for futur epochs (not yet in TRL)
```

Comment (pramodith) [reply docs/source/paper_index.md]: Oh that's a nice list! I might have a PR for `batch definition` coming up, have a draft version that needs polishing.

File: trl/trainer/grpo_config.py (hunk 7 above)
Comment (pramodith) [reply trl/trainer/grpo_config.py]: Done

File: trl/trainer/grpo_config.py [hunk 10]
```diff
@@ -166,6 +166,8 @@ class GRPOConfig(TrainingArguments):
         epsilon_high (`float`, *optional*):
             Upper-bound epsilon value for clipping. If not specified, it defaults to the same value as the lower-bound
             specified in argument `epsilon`. Paper [DAPO](https://huggingface.co/papers/2503.14476) recommends `0.28`.
+            When used with `loss_type='cispo`, this corresponds to the ε_max param specified in the
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
            When used with `loss_type="cispo"`, this corresponds to the ε_max param specified in the
```

File: trl/trainer/grpo_config.py [hunk 11]
```diff
@@ -166,6 +166,8 @@ class GRPOConfig(TrainingArguments):
...
+            [ScaleRL paper](https://arxiv.org/pdf/2510.13786) and the recommended value is `5.0`.
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
            [ScaleRL paper](https://huggingface.co/papers/2510.13786) and the recommended value is `5.0`.
```

File: trl/trainer/grpo_config.py [hunk 12]
```diff
@@ -201,6 +203,11 @@ class GRPOConfig(TrainingArguments):
               batch. Note that normalization is performed over the local batch only, so results may slightly vary
               depending on the local batch size, despite a constant effective batch size. When using
               `per_device_train_batch_size==1`, the loss is equivalent to the GRPO loss.
+            - `"cispo"`: Clips the importance sampling weights instead of the advantage scaled importance weights.
+              The clipped weights are then multiplied with the advantages and policy model's log probs.
+              Individual token losses are aggregated by normalizing with the number of active tokens in
+              the global accumulated batch. This method was introduced in the
+              [MiniMax-M1 paper][MiniMax-M1 paper](https://huggingface.co/papers/2506.13585).
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
              [MiniMax-M1 paper](https://huggingface.co/papers/2506.13585).
```

File: trl/trainer/grpo_config.py [hunk 13]
```diff
@@ -533,7 +540,9 @@ class GRPOConfig(TrainingArguments):
         default=None,
         metadata={
             "help": "Upper-bound epsilon value for clipping. If not specified, it defaults to the same value as the "
-            "lower-bound specified in argument `epsilon`. Paper DAPO recommends `0.28`."
+            "lower-bound specified in argument `epsilon`. Paper DAPO recommends `0.28`. "
+            "When used with `loss_type='cispo`, this corresponds to the ε_max param specified in the"
+            "[ScaleRL paper](https://arxiv.org/pdf/2510.13786) and the recommended value is `5.0`."
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
            "[ScaleRL paper](https://huggingface.co/papers/2510.13786) and the recommended value is `5.0`."
```

File: trl/trainer/grpo_config.py [hunk 14]
```diff
@@ -582,6 +591,11 @@ class GRPOConfig(TrainingArguments):
             "Note that normalization is performed over the local batch only, so results may slightly vary depending "
             "on the local batch size, despite a constant effective batch size. When using "
             "`per_device_train_batch_size==1`, the loss is equivalent to the GRPO loss."
+            "'cispo': Clips the importance sampling weights instead of the advantage scaled importance weights. "
+            "The clipped weights are then multiplied with the advantages and policy model's log probs. "
+            "Individual token losses are aggregated by normalizing with the number of active tokens in "
+            "the global accumulated batch. This method was introduced in the "
+            "[MiniMax-M1 paper][MiniMax-M1 paper](https://huggingface.co/papers/2506.13585)."
```
Comment (qgallouedec) [inline trl/trainer/grpo_config.py]:
```suggestion
            "[MiniMax-M1 paper](https://huggingface.co/papers/2506.13585)."
```

**Comment by pramodith** [comment]
Failing CI seems unrelated to this PR, going ahead and merging.

**Comment by Neelectric** [comment]
In the [Paper Index docs](https://huggingface.co/docs/trl/main/en/paper_index#the-art-of-scaling-reinforcement-learning), this is mentioned as a code snippet:

```
from trl import GRPOConfig

config = GRPOConfig(
    loss_type="cispo",
    epsilon_high=5.0,
    num_completions=16,
    scale_rewards="batch",
    cast_lm_head_to_fp32=True
)
```
I believe it should be `num_generations` instead of `num_completions`?

**Comment by qgallouedec** [comment]
True! Do you want to open a PR?

**Comment by Neelectric** [comment]
Fair point! Just saw that @pramodith [beat me to it ](https://github.com/huggingface/trl/pull/4515) :)
